### ISSUE TITLE
Bug & Feature in ContextAwareRandomizer class

### ISSUE THREAD

**Issue by nrenzoni** [j-easy/easy-random#356]
Feature Request (see comment below in `getRandomValue()`):

```java
import org.jeasy.random.*;
import org.jeasy.random.api.*;
import org.junit.*;
import java.lang.annotation.*;
import static java.lang.annotation.ElementType.FIELD;
import static java.lang.annotation.RetentionPolicy.RUNTIME;

public class Tester {

    private static EasyRandom easyRandom;

    @Before
    public void setup() {
        EasyRandomParameters parameters = new EasyRandomParameters()
                .randomize(FieldPredicates.isAnnotatedWith(ExampleAnnotation.class), new CustomRandomizerContext());
        easyRandom = new EasyRandom(parameters);
    }

    @Test
    public void test() {
        A a = easyRandom.nextObject(A.class);
    }

    static class CustomRandomizerContext implements ContextAwareRandomizer<D> {

        RandomizerContext context;

        @Override
        public void setRandomizerContext(RandomizerContext context) {
            this.context = context;
        }

        @Override
        public D getRandomValue() {
            /*
             Problem: want to access "some value" (see below) of annotation here; need a reference to Field (like below) of the object, or to the enclosing object B b.
            */
            D d = new D();
            d.value = context.getField().getAnnotation(ExampleAnnotation.class).value();
            return d;
        }
    }

    static class A {
        private B b;
    }

    static class B {
        private C c;

        @ExampleAnnotation("some value")
        private D d;
    }

    static class C {
    }

    static class D {
        private String name;
    }

    @Target({FIELD})
    @Retention(RUNTIME)
    @Documented
    public @interface ExampleAnnotation {
        String value();
    }
}
```

**Comment by nrenzoni** [comment]
Bug found in `RandomizerContext.getCurrentObject()`, where not always referencing the current object.
- Since `getCurrentObject()` implementation in `RandomizationContext` returns `randomizedObject`, and this variable is only assigned in `setRandomizedObject()`, which is only called from `doPopulateBean()`, but not in all cases. So the correct way to do implement `getCurrentObject()` would be to use `stack` in `RandomizationContext`, and to retrieve the last element from the `stack` in the case where there are elements, or if the `stack` is empty, to return the root element.

**Comment by nrenzoni** [comment]
I pushed 2 commits to address these 2 respective issues: [branch in my fork](https://github.com/nrenzoni/easy-random/commits/RandomizerContext_bug_fixes_and_features).

**Comment by benas** [comment]
#### In regards to the bug

I was able to reproduce the issue. Indeed, `RandomizationContext#getCurrentObject` is incorrect in the above case, it returns an instance of `C` where it should return an instance of the enclosing `B`.

@nrenzoni Your fix LGTM  so I applied it. The credit goes to you so thank you! I added you to the contributors list. The fix is deployed in `4.1.0-SNAPSHOT` if you want to give it a try already. I will release 4.1.0 in the next few days.

#### In regards to the requested feature

The requested feature asks to expose the current `Field` in the context. I believe this is not required now that the bug is fixed and we can correctly access to the enclosing object.

> need a reference to Field (like below) of the object, or to the enclosing object B b.

The combination of the full path to the current field (in dotted notation) + the enclosing object is enough to get the field. An example based on the requirement above can be found [here](https://github.com/j-easy/easy-random/blob/master/easy-random-core/src/test/java/org/jeasy/random/RandomizationContextTest.java#L233-L242). It requires to extract the field name from the full path but that is not an issue IMO.

**Comment by nrenzoni** [comment]
Thank you for taking care of this @benas
